### PR TITLE
added research file

### LINKED ISSUE #6: Research on Fine-Tuning Process for CodeT5 using Bioinformatics Datasets
Research and document the process of fine-tuning the CodeT5 model using bioinformatics datasets. The objective is to outline the required steps, including dataset preprocessing, training setup, hyperparameter tuning, and evaluation.

Acceptance Criteria:
- Outline the step-by-step fine-tuning process for CodeT5 and the code that we will be using in each step.
- Include dataset preparation steps (cleaning, formatting, tokenization) with the codes.
- Describe evaluation methods and expected metrics for fine-tuned model performance.
- Document findings in a Markdown (.md) file.

### PR DESCRIPTION
**Pull request by Omp06**
Fixes #6 

**What was changed?**

I created a markdown file that lists the steps to fine-tune the Codet5 model. I focused on how to preprocess the data, set up the training, hyperparameter tuning, and evaluation of the model. I also gave code snippets that show a very general example of what the code will look like in each step. 

**Why was it changed?**

This research will allow us to take the Codet5 model and adjust it to how we want it to behave. We can take the datasets we have found and train the Codet5 model. These training steps will help to make the Codet5 model more robust. 

**How was it changed?**

The markdown file was named Fine_Tuning_Research.md. I wrote an overview of what the document will hold. Following this,  I described the data collection and cleaning as well as formatting and tokenization steps that need to be taken in order for the model to be utilized correctly. I also give code on how the model will be initialized and the training arguments that should be sent to the model when executed. This was detailed in code snippets. I also described how to use Optuna for hyperparameter optimization using code as well as how to evaluate it after the parameters have been set. 
*Here, get into detail about what files you modified, and talk about the most important lines in regards to fixing the issue.*

### CODE REVIEW

File: Fine_Tuning_Research.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+#### Example Code for Cleaning:
+
+For python code: 
+
+```python
+import re
+
+def clean_code(code_snippet):
+    # Remove comments
```
Comment (oam67) [inline Fine_Tuning_Research.md]: Would we want to maintain certain comments in the code datasets since CodeT5 works with code paired with comments? Maybe we can keep comments that are relevant to code explanation and/or logic explanation that supports the understanding of the code.

File: Fine_Tuning_Research.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+# Print sequence
+print(seq)
+"
+cleaned_code <- clean_bioinformatics_code(raw_code)
+cat(cleaned_code)
+```
+
+### 1.3 Formatting and Tokenization
+- CodeT5 uses a code-specific **BPE tokenizer** trained using the Hugging Face `RobertaTokenizer`.
```
Comment (oam67) [inline Fine_Tuning_Research.md]: Will we be using RobertaTokenizer for the bioinformatics use case or would we have to work on a new tokenizer that recognizes common bioinformatics functions and libraries? How can we ensure correct tokenization before training?

File: Fine_Tuning_Research.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    return result.training_loss
+
+study = optuna.create_study(direction="minimize")
+study.optimize(objective, n_trials=10)
+print("Best hyperparameters:", study.best_params)
+```
+---
+
+## 4. Evaluation
+### 4.1 Metrics
```
Comment (oam67) [inline Fine_Tuning_Research.md]: Maybe we can also look at the metrics used by BioCoder to evaluate and benchmark the fine-tuned CodeT5 model. It would be good to compare how our fine tuned model performs in comparison to the other general-purpose models.

File: Fine_Tuning_Research.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import evaluate
+
+metric = evaluate.load("bleu")
+
+def compute_metrics(eval_pred):
+    predictions, labels = eval_pred
+    decoded_preds = tokenizer.batch_decode(predictions, skip_special_tokens=True)
+    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
+    return metric.compute(predictions=decoded_preds, references=decoded_labels)
+```
```
Comment (oam67) [inline Fine_Tuning_Research.md]: Overall great work! This information will be helpful when we begin the fine-tuning and evaluation process!

**Review by oam67 — COMMENTED** [review]
Submitting code review.

**Review by AjithAkuthota23 — CHANGES_REQUESTED** [review]
Good work. I know it looks vague right now but it will make more sense once we have the actual data that we need to work on.


Can you create a new folder saying Fine Tuning and add the file to that?

**Review by AjithAkuthota23 — APPROVED** [review]
